draft-ietf-quic-recovery.md

```diff
@@ -398,8 +417,80 @@ A packet sent on an RTO alarm MUST NOT be blocked by the sender's congestion
+handshake latency.
+
+### Empty ACK frames
+
+The EMPTY_ACK frame enables faster recovery of lost Initial and Handshake
+packets when the only other outstanding packets are undecryptable.
+In that case, the peer can’t decrypt the packet or packet number,
+but they do know a packet for a QUIC connection of a supported version was
+received.
+
+The receiver SHOULD send an EMPTY_ACK frame soon (e.g., 1ms) after
+undecryptable packets are received, even if those received packets are
+not buffered for later decryption.  The small delay allows for cases when
+0-RTT packets are reordered in front of the Initial, which is not uncommon
+on networks that prioritize small packets.  The receiver should limit the
+number of EMPTY_ACK frames sent to one per packet number space per RTT.
+
+When an EMPTY_ACK frame is received, a sender should immediately
+retransmit the missing handshake packet(s) as though the handshake timer
+fired and re-arm the handshake timer when the handshake packets are sent.
+If the missing handshake data was timer retransmitted after the packets
+that triggered the empty ack, then the empty ack should be ignored.
+
+An EMPTY_ACK frame does not acknowledge a new packet, and in cases
+when multiple packets are outstanding, the RTT signal is ambiguous,
+so it should not be used like an RTT signal from a newly acknowledged
+packet.  It MAY change the connection’s default RTT if no RTT measurements
+have been taken.
+
+### Coalesced Packets
+
+Despite loss recovery being separate for different packet number spaces,
+the ability to send a coalesced packet enables faster recovery with small,
+and sometimes no overhead.  The acknowledgement of a coalesced packet
+allows QUIC recovery to use early retransmit to determine if any prior
+packets in that space were lost without waiting for timeouts.
+
+This optimization is particularly useful when:
+
+ * Retransmitting the client’s Initial, which must be padded to a full
+   sized packet, so the datagram typically has extra space to retransmit
+   some outstanding 0-RTT data.
+ * The server sends 1-RTT data soon after the final handshake flight
+   (containing the server Finished) and can proactively retransmit an empty
+   CRYPTO_HS frame bundled with one or more 1-RTT packets.
+ * The clients sends 1-RTT data soon after the final TLS flight
+   (containing the client Finished) and can proactively retransmit the
+   final client flight with one or more 1-RTT packets.
+
+### Implicit Acknowledgements
+
+Handshake data may be cancelled by handshake state transitions.
+
+In particular:
+
+ * A peer processing data in a HANDSHAKE packet indicates
+   the INITIAL packet(s) have been delivered.
+ * A peer processing 1-RTT packets indicates all CRYPTO_HS data in
+   HANDSHAKE packets has been delivered.
+
 ## Generating Acknowledgements
 
+An ACK frame acknowledges packets from only one packet number space.
+It may be optimal to store multiple ACK frames at once until the
```

Comment:
Where do we store ACK frames? This sounds like an implementation detail.
In quic-go, I never store ACK frames, but just a data structure containing the ranges of received packets, and only generate the ACK frame data structure when actually sending one.